Specify a DialContext in storage plugin clients

mattcary opened the pull request:
Adds filtering of hosts to DialContexts. the provided DialContext wraps existing
clients' DialContext in an attempt to preserve any existing timeout
configuration. In some cases, we may replace infinite timeouts with golang
defaults.

- scaleio: tcp connect/keepalive values changed from 0/15 to 30/30
- storageos: no change

The added unit tests will fail to compile if the vendor patches are
accidentally reverted.

gluster and quobyte should be updated, but that is blocked on upstream changes on their drivers.

/kind bug
**Which issue(s) this PR fixes**:
#91542

```release-note
kube-controller-manager: volume plugins can be restricted from contacting local and loopback addresses by setting `--volume-host-allow-local-loopback=false`, or from contacting specific CIDR ranges by setting `--volume-host-cidr-denylist` (for example, `--volume-host-cidr-denylist=127.0.0.1/28,feed::/16`)
```

mattcary commented:
/cc msau
/cc tallclair

mattcary commented:
Note that the current behavior of this patch is to always deny dialing localhost and similar addresses (as defined by proxy/util:IsProxyableIP).

It's not clear to me that's the behavior we want and I'd appreciate any opinions from the community.

(in addition to opinions about everything else in the CL :)

msau42 commented on pkg/volume/BUILD:
```diff
@@ -54,6 +55,7 @@ go_library(
 go_test(
     name = "go_default_test",
     srcs = [
+        "gke_patch_test.go",
```
Remove this

mattcary replied:
done.

Jonathon suggested a different test to replace it but I'm not clear on what we're shooting for here.

msau42 commented on vendor/github.com/heketi/heketi/client/api/go-client/client.go:
```diff
@@ -14,13 +14,15 @@ package client
 
 import (
 	"bytes"
+	"context"
```
We need to update vendor with the upstream changes instead of changing vendor directly. cc @misterikkit

mattcary replied:
Thanks Michelle & Jonathon.

I'm not familiar with how to update vendor/, any pointers would be appreciated (here or offline).

msau42 commented:
/assign @misterikkit @jsafrane @gnufied 
/ok-to-test

tallclair commented:
It looks like this PR depends on https://github.com/heketi/heketi/pull/1707? I'd prefer to not block on https://github.com/heketi/heketi/pull/1707 merging. In that case, I'd say lets drop the update to GlusterFS, and follow up with another PR once the dependency merges.

msau42 commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -818,6 +819,11 @@ func (adc *attachDetachController) GetSubpather() subpath.Interface {
 }
 
+func (adc *attachDetachController) GetFilteredDialOptions() *proxyutil.FilteredDialOptions {
+	// FilteredDial not needed in attachdetach controller
```
If the targeted plugins implement Attacher, then we do need it.

mattcary replied:
Done.

msau42 commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -447,3 +448,9 @@ func (expc *expandController) GetSubpather() subpath.Interface {
+
+
+func (adc *expandController) GetFilteredDialOptions() *proxyutil.FilteredDialOptions {
+	// not needed for expand controller
```
same thing if the plugins implement Expand

mattcary replied:
Done

msau42 commented on cmd/kube-controller-manager/app/options/persistentvolumebindercontroller.go:
```diff
@@ -63,6 +69,17 @@ func (o *PersistentVolumeBinderControllerOptions) Validate() []error {
+
 	errs := []error{}
+	for _, cidr := range strings.Split(cleanedList, ",") {
+		if _, _, err := net.ParseCIDR(strings.TrimSpace(cidr)); err != nil {
```
Can we validate this at a higher layer? especially if we need to support multiple volume controllers?

mattcary replied:
The string is already validated at cmd/kube-controller-manager/app/options/persistentvolumebindercontroller.go:67, is that what you were thinking?

msau42 commented on pkg/proxy/util/utils.go:
```diff
@@ -311,3 +322,44 @@ func EnsureSysctl(sysctl utilsysctl.Interface, name string, newVal int) error {
+		}
+
+		for _, host := range resp {
+			if err := isProxyableIP(host.IP); err != nil{
```
@jsafrane @gnufied would like your thoughts on this, if always disallowing localhost is going to break users.

jsafrane replied:
It will break some of my test deployments with local-up-cluster.sh. I am not sure about real users though.

mattcary replied:
This is to mitigate the problem described here: https://groups.google.com/g/kubernetes-security-announce/c/kEK27tqqs30

How important is it for local-up-cluster.sh to work with localhost? Are there workarounds?

msau42 replied:
Since we want to backport this fix to older releases, I would be more cautious about changing default behavior that we think has the potential to break someone. We try to minimize changes for patch releases (and offer a way to dsiable it), but we can consider changing things in a new minor release after a deprecation period.

mattcary replied:
Yeah, that makes sense. On the other hand, secure-by-default is a good stance. If the only impacted scenarios are testing and similar, and it can be fixed with a flag, would that be enough? 

@jsafrane: if we could disable the IsProxyable check through the command-line flag, would that be enough?

Proposal: if --dial-host-cidr-denylist=none, then we do no filtering. If it's empty or contains CIDR ranges, then we do IsProxyable in addition to any ranges passed.

jsafrane replied:
`--dial-host-cidr-denylist=none`: LGTM
@humblec, AFAIK our production Gluster / Heketi never works on 127.0.0.1, it always run in a container that has  a real IP address, right?

humblec replied:
@jsafrane yeah, the heketi server is running in a pod with an IP, what SC carry for contacting heketi service is a heketi service url on top of this.

humblec replied:
@mattcary as @msau42 and @jsafrane suggested, having a fallback option with a knob/flag like `secure-be-default` should help. Maybe some scenarios/usages are unknown/missed from our radar and cause the breakage due to this change.

tallclair replied:
+1 for `--dial-host-cidr-denylist=none`. Let's make that the default behavior for the flag in this PR so it's opt-in for the cherry-picks, and then open a new PR to make it opt out (not for cherry-picking).

fedebongio commented:
/cc @cheftako

mattcary commented:
Makes sense Tim. Will do.

> It looks like this PR depends on [heketi/heketi#1707](https://github.com/heketi/heketi/pull/1707)? I'd prefer to not block on [heketi/heketi#1707](https://github.com/heketi/heketi/pull/1707) merging. In that case, I'd say lets drop the update to GlusterFS, and follow up with another PR once the dependency merges.

mattcary committed:
Adds filtering of hosts to DialContexts.

The provided DialContext wraps existing clients' DialContext in an attempt to
preserve any existing timeout configuration. In some cases, we may replace
infinite timeouts with golang defaults.

- scaleio: tcp connect/keepalive values changed from 0/15 to 30/30
- storageos: no change

mattcary committed:
update storageos vendor for FilteredDial change

mattcary commented:
I've removed gluster and updated vendor/ according to the instructions at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md.

I've kept the vendor/ change as a separate commit for clarity but can squash them if desired.

Does anyone have any comments on my proposal? 

If --dial-host-cidr-denylist=none, then we do no filtering. If it's empty or contains CIDR ranges, then we do IsProxyable in addition to any ranges passed.

humblec commented:
> I've removed gluster and updated vendor/ according to the instructions at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md.
> 
> I've kept the vendor/ change as a separate commit for clarity but can squash them if desired.

@mattcary i dont see above mentioned commit in this PR. 
>

mattcary commented:
> > I've removed gluster and updated vendor/ according to the instructions at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md.
> > I've kept the vendor/ change as a separate commit for clarity but can squash them if desired.
> 
> @mattcary i dont see above mentioned commit in this PR.
> 
> >

https://github.com/kubernetes/kubernetes/pull/91785/commits/1485774b35d2e0763f7659b90422c544c4951361

tallclair commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -632,3 +649,15 @@ func getNodeCIDRMaskSizes(clusterCIDRs []*net.IPNet, maskSizeIPv4, maskSizeIPv6
 	return nodeMaskCIDRs
 }
+
+func parseControllerDialhostIPDenylist(ctx *ControllerContext) ([]*net.IPNet, error) {
```
nit: capitalize host
```suggestion
func parseControllerDialHostIPDenylist(ctx *ControllerContext) ([]*net.IPNet, error) {
```

mattcary replied:
done

tallclair commented on cmd/kube-controller-manager/app/options/persistentvolumebindercontroller.go:
```diff
@@ -63,6 +69,17 @@ func (o *PersistentVolumeBinderControllerOptions) Validate() []error {
 		return nil
 	}
 
+	cleanedList := strings.TrimSpace(o.DialHostCIDRDenylist)
```
The validation logic should be consistent with `parseControllerDialhostIPDenylist`, so I don't think you should trim spaces here. Alternatively, you could refactor the parse function to take the string argument directly, and just call it here as well (check for the error, and through away the parsed list).

mattcary replied:
Good point. Refactored to use netutils.ParseCIDRs which I think accomplishes your suggestion.

mattcary replied:
Actually, with the "none" change suggested below I ended up doing exactly your refactoring.

tallclair commented on pkg/controller/volume/attachdetach/attach_detach_controller_test.go:
```diff
@@ -262,6 +263,30 @@ func attachDetachRecoveryTestCase(t *testing.T, extraPods1 []*v1.Pod, extraPods2
+		false,
+		1*time.Second,
+		DefaultTimerConfig,
+		nil, /* filteredDialOptions */
```
I think you need to add this arg at line 177 too? How come you need to create a new controller here?

mattcary replied:
Not sure what happened there. The diff is no reasonable.

mattcary replied:
Sorry, I tried to write "fixed, the diff is now reasonable".

A new controller didn't need to be made, maybe there was a rebase problem.

tallclair commented on pkg/proxy/util/utils.go:
```diff
@@ -311,3 +322,44 @@ func EnsureSysctl(sysctl utilsysctl.Interface, name string, newVal int) error {
+			}
+			if opts != nil && opts.DialHostCIDRDenylist != nil {
+				if err := IsAllowedHost(ctx, host.IP, opts.DialHostCIDRDenylist); err != nil {
+				return nil, err
```
nit: indentation (`hack/update-gofmt.sh` should fix it?)

mattcary replied:
w/rt --dial-host-cidr-denylist=none, your idea is to make "none" the default value for the flag, so that to make it opt-out the only change will be for the flag default?

mattcary commented:
Opting in is now set via a flag default.

I'm not sure of a good way to test that that's actually plumbed through in an e2e manner, any suggestions are welcome. I suppose any e2e tests that rely on it will fail.

tallclair commented:
Yeah, I'm not sure this needs an E2E test. I think unit tests & manual testing are the way to go here.

cheftako commented:
Looks like assorted lint and static check errors.

mattcary commented:
> Looks like assorted lint and static check errors.

Yeah, more significantly I think I've figured out I did something wrong with changing pkg/controller/volume/persistentvolume/config, the tests seem to be failing because of some v1alpha1=>v1 type conversion.

I'll ask around as I'm not figuring it out on my own, but if anyone has any hints I'll take them!

mattcary commented:
> > Looks like assorted lint and static check errors.
> 
> Yeah, more significantly I think I've figured out I did something wrong with changing pkg/controller/volume/persistentvolume/config, the tests seem to be failing because of some v1alpha1=>v1 type conversion.
> 
> I'll ask around as I'm not figuring it out on my own, but if anyone has any hints I'll take them!

With some help (thanks Daniel!) I figured out that the kube-controller-manager config in staging was getting auto-regenerated correctly. We'll see if that moves the test forward.

cheftako commented on pkg/controller/volume/persistentvolume/config/types.go:
```diff
@@ -28,6 +28,9 @@ type PersistentVolumeBinderControllerConfiguration struct {
 	PVClaimBinderSyncPeriod metav1.Duration
 	// volumeConfiguration holds configuration for volume related features.
 	VolumeConfiguration VolumeConfiguration
+	// DialHostCIDRDenylist is a comma-separated list of IPs that should not be dialed by the
```
Your previous comments and examples show this as a list of CIDRs not IPs. The name also suggests this should be CIDRs.

mattcary replied:
Done, thank you for catching.

cheftako commented on staging/src/k8s.io/kube-controller-manager/config/v1alpha1/types.go:
```diff
@@ -427,6 +427,9 @@ type PersistentVolumeBinderControllerConfiguration struct {
 	PVClaimBinderSyncPeriod metav1.Duration
 	// volumeConfiguration holds configuration for volume related features.
 	VolumeConfiguration VolumeConfiguration
+	// DialHostCIDRDenylist is a comma-separated list of IPs that should not be dialed by the
```
Ditto IPs.

mattcary replied:
And ditto :)

mattcary commented:
I think I figured out some of my presubmit woes: I have go 1.14.4 locally and I think that doesn't play nice with some of the tools. Trying to resolve it.

mattcary commented:
/retest

mattcary commented:
/retest

cheftako commented:
Run ./hack/update-bazel.sh to fix verify

mattcary commented:
> Run ./hack/update-bazel.sh to fix verify

Thanks. I had run that already but I guess it got deranged.

mattcary commented:
/retest

Integration seems to be flaky

msau42 commented on vendor/github.com/quobyte/api/rpc_client.go:
```diff
@@ -112,7 +113,17 @@ func (client QuobyteClient) sendRequest(method string, request interface{}, resp
+		if err != nil {
+			return (err)
+		}
+		log.Printf("Warning: Dumping full reply body:\n%s\n", string(body))
```
@tallclair wdyt, this basically undoes the logging change?

tallclair replied:
oh yeah, can we not do this? Techincally the new dial restrictions should mitigate it, but I'd rather keep the full response out of the logs.

misterikkit replied:
https://github.com/quobyte/api/pull/20

msau42 commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -323,6 +329,11 @@ func startAttachDetachController(ctx ControllerContext) (http.Handler, bool, err
 		return nil, true, fmt.Errorf("failed to probe volume plugins when starting attach/detach controller: %v", err)
 	}
 
+	filteredDialOptions, err := options.ParseDialHostDenyList(ctx.ComponentConfig.PersistentVolumeBinderController.DialHostCIDRDenylist)
```
Is it odd that that the options for the other volume controllers come from the PV controller config? Should we have a separate "VolumeControllers" config that applies to all?

mattcary replied:
That's true, I had set it up this way when it was only being used in the controller manager.

However I confess to not really understanding how these configurations are used and what the implications for how they're organized are.

liggitt replied:
> I had set it up this way when it was only being used in the controller manager.

is this used in other components? I'm not seeing flags or config that would enable that

cheftako commented on pkg/proxy/util/utils.go:
```diff
@@ -124,6 +125,16 @@ func IsProxyableHostname(ctx context.Context, resolv Resolver, hostname string)
 }
 
+// IsAllowedHost checks if the given IP host address is in a network in the denied list.
+func IsAllowedHost(ctx context.Context, host net.IP, denied []*net.IPNet) error {
```
Small quibble but I expect any method with an Is prefix to return a bool

mattcary replied:
I don't disagree, but this is named & used in the same way as IsProxyableHostname which also returns an error rather than a bool. I think it's better to be consistent; the principled alternative would be to change IsProxyableHostname as well which is probably excessive.

cheftako commented:
Looks fine to me. Deferring to @msau42

msau42 commented:
/lgtm
/assign @liggitt

mattcary commented:
FYI quobyte has been removed in the recent push.

mattcary commented:
/restest

Failures seem to be timeouts; the verify job passed all tests, has a panic in the build log but the associated tests all report passing, as near as I can tell by looking through the 16M of noise in the log. Sigh.

mattcary commented:
/retest

mattcary commented:
/test pull-kubernetes-verify

I'm not sure what's going on with the verify. The job history shows that most other tasks are failing in a similar way (all tests pass but for some reason the job fails). The one passing job has the same panics in the logs and yet passes...

liggitt commented on pkg/proxy/util/utils.go:
```diff
@@ -311,3 +322,54 @@ func EnsureSysctl(sysctl utilsysctl.Interface, name string, newVal int) error {
+}
+
+// NewFilteredDialContext returns a DialContext function that filters connections based on a FilteredDialOptions.
+func NewFilteredDialContext(wrapped DialContext, opts *FilteredDialOptions) DialContext {
```
take a `resolv Resolver` rather than hard-coding to `net.DefaultResolver`? will allow easier unit testing

mattcary replied:
Done

liggitt commented on cmd/kube-controller-manager/app/options/persistentvolumebindercontroller.go:
```diff
@@ -43,6 +48,8 @@ func (o *PersistentVolumeBinderControllerOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&o.VolumeConfiguration.EnableHostPathProvisioning, "enable-hostpath-provisioner", o.VolumeConfiguration.EnableHostPathProvisioning, "Enable HostPath PV provisioning when running without a cloud provider. This allows testing and development of provisioning features.  HostPath provisioning is not supported in any way, won't work in a multi-node cluster, and should not be used for anythi ...
 	fs.BoolVar(&o.VolumeConfiguration.EnableDynamicProvisioning, "enable-dynamic-provisioning", o.VolumeConfiguration.EnableDynamicProvisioning, "Enable dynamic provisioning for environments that support it.")
 	fs.StringVar(&o.VolumeConfiguration.FlexVolumePluginDir, "flex-volume-plugin-dir", o.VolumeConfiguration.FlexVolumePluginDir, "Full path of the directory in which the flex volume plugin should search for additional third party volume plugins.")
+	fs.StringVar(&o.DialHostCIDRDenylist, "dial-host-cidr-denylist", o.DialHostCIDRDenylist, "A comma-separated list of CIDR ranges to avoid dialing from volume plugins, or `none'. If `none', nothing is filtered. Otherwise, non-proxyable hosts are filtered in addition to any specified CIDR ranges.")
```
slight preference for making `DialHostCIDRDenylist` a `[]string` and using `StringSliceVar` which handles the comma-separated parsing for you

mattcary replied:
Using it as a string makes it slightly cleaner to parse the "none" option, which shouldn't be one option of many. Conceptually, this is a flag of type {none, comma-separated-CIDRs} so it isn't exactly a StringSliceVar.

liggitt replied:
If I understand correctly, we're logically populating two things in the filteroptions struct:
* DenyNonproxyableHosts bool
* DialHostCIDRDenylist []string

For backports, we want to default those to `false, []`

Going forward, we want to default those to `true, []`

Trying to address both of those values in a single flag is pretty confusing. If the `DenyNonproxyableHosts` field was tied to its own flag (which defaulted to `false` in backports and `true` going forward), would that be easier to reason about? That would let the CIDR deny-list be a simple list with no special "none" semantics.

mattcary replied:
Yeah, that's a good idea. Done.

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -76,13 +78,17 @@ type apiImplementer interface {
 		return nil
 	}
+	if u.host == nil {
+		return nil
```
returning nil here changes the behavior of anything currently using storageosUtil... I would expect this to error or continue with unfiltered dials

liggitt commented on cmd/kube-controller-manager/app/options/persistentvolumebindercontroller.go:
```diff
@@ -64,5 +72,28 @@ func (o *PersistentVolumeBinderControllerOptions) Validate() []error {
+// FilteredDialOptions that block non-proxyable IPs along with any CIDR ranges
+// in arg is returned.
+func ParseDialHostDenyList(arg string) (*proxyutil.FilteredDialOptions, error) {
+	if arg == "none" {
+		return &proxyutil.FilteredDialOptions{DialHostCIDRDenylist: nil, DenyNonproxyableHosts: false}, nil
+	}
+	if len(arg) == 0 {
+		return &proxyutil.FilteredDialOptions{DialHostCIDRDenylist: nil, DenyNonproxyableHosts: true}, nil
+	}
```
the distinction between these is pretty hard to follow, especially when combined with the default flag/configfile value

possible behaviors:
1. anything allowed (current behavior)
2. deny nonproxyable hosts
3. deny specific list of CIDRs
4. deny nonproxyable hosts and specific list of CIDRs

is it important to be able to be able to achieve all four of these behaviors?

possible flag-configured values:
1. no flag
2. flag with no value (`--dial-host-cidr-denylist`)
3. flag with explicit `none` value (`--dial-host-cidr-denylist=none`)
4. flag with cidr values (`--dial-host-cidr-denylist=1.2.3.4/16,...`)

flag option 2 is confusing to me

possible config file values:
1. no field
2. field with explicit null or empty list (`DialHostCIDRDenylist: null` or `DialHostCIDRDenylist: []`)
3. field with explicit `none` (`DialHostCIDRDenylist: ["none"]`)
4. field with explicit cidr values (`DialHostCIDRDenylist: ["1.2.3.4/16",...]`)

config options 1 and 2 would be treated the same by defaulting case and turned into `["none"]`... distinguishing between missing, nil, and empty lists has caused lots of confusion in the past, so we avoid it now

liggitt replied:
I'm actually having a hard time telling what the default behavior is if someone invokes kube-controller-manager with no flag or config option provided for this... does this PR switch the default to `DenyNonproxyableHosts=true`?

mattcary replied:
w/rt using in other components: there is no flag, but the plumbing is there, see eg pkg/controller/volume/attachdetach/attach_detach_controller.go.

w/rt flag options, it is confusing and I apologize. The behaviors we want are:

1) anything allowed (current behavior)
2) IsProxyable-like behavior onky. This can't be easily specified with a CIDR list without manually enumerating what ranges, eg, localhost map to for both IPv4 and IPv6
3) Custom CIDR ranges + IsProxyable behavior.

See ParseDialHostDenyList in cmd/kube-controller-manager/app/options/persistentvolumebindercontroller.go where things are split up.

mattcary commented:
@Jordan: Somehow github is not letting me reply to some of your comments, probably because of a force push? Anyway...

Regarding the default flag value, I'm not sure how to detect in defaults.go if a bool value is unset. Or should I just unconditionally set it and let it be overridden by the flag?

Regarding the ctx in IsAllowedHost, thanks that was left over from when we resolved in that ftn

regarding the NewAPI call in storageos_util, done, I think I mistakenly thought returning nil would signal an error.

Regarding use in other components, it currently isn't wired to a flag but the internal plumbing was put by a suggestion of msau@'s, see pkg/controller/volume/attachdetach/attach_detach_controller.go.

With the addition of the new bool flag I hope your concerns about the flag parsing are now resolved.

liggitt commented on pkg/controller/volume/persistentvolume/config/v1alpha1/zz_generated.conversion.go:
```diff
@@ -105,6 +105,8 @@ func autoConvert_v1alpha1_PersistentVolumeBinderControllerConfiguration_To_confi
 	if err := Convert_v1alpha1_VolumeConfiguration_To_config_VolumeConfiguration(&in.VolumeConfiguration, &out.VolumeConfiguration, s); err != nil {
 		return err
 	}
+	// WARNING: in.DialHostCIDRDenylist requires manual conversion: inconvertible types (*[]string vs []string)
```
making both the internal and external versions use `[]string` will fix this

liggitt commented on pkg/controller/volume/persistentvolume/config/v1alpha1/zz_generated.conversion.go:
```diff
@@ -113,6 +115,8 @@ func autoConvert_config_PersistentVolumeBinderControllerConfiguration_To_v1alpha
 	if err := Convert_config_VolumeConfiguration_To_v1alpha1_VolumeConfiguration(&in.VolumeConfiguration, &out.VolumeConfiguration, s); err != nil {
 		return err
 	}
+	// WARNING: in.DialHostCIDRDenylist requires manual conversion: inconvertible types ([]string vs *[]string)
```
making both the internal and external versions use `[]string` will fix this

liggitt commented on staging/src/k8s.io/kube-controller-manager/config/v1alpha1/types.go:
```diff
@@ -427,6 +427,12 @@ type PersistentVolumeBinderControllerConfiguration struct {
+	DialHostCIDRDenylist *[]string
+	// DialHostDenyNonproxyable indicates if nonproxyable hosts should be denied
+	// when dialing from plugins.
+	DialHostDenyNonproxyable bool
```
also, is there a better name than `DialHostDenyNonproxyable`? We don't proxy to volume plugin hostnames, so it's a bit odd to make the config/flag call out "nonproxyable" addresses. Is that a term of art that users will know means loopback + link-local-unicast + link-local-multicast + interface-local-multicast?

cc @thockin in case there's a nice standard term for this set of IP ranges.

might be easier to understand what the option is doing if it is opt in and defaults false (in 1.19+), something like:
```
VolumeHostAllowLocalLoopback *bool
```

mattcary commented:
@jordan: flag/option names changed to use VolumeHost as you suggested, changed direction of the bool, used the bool*, fixed the minor type issues.

mattcary commented:
/test pull-kubernetes-e2e-kind

Failed on timeout which I suspect is a flake.

mattcary commented:
/test pull-kubernetes-e2e-gce-100-performance

More timeout flakiness

mattcary commented:
/test pull-kubernetes-e2e-gce-100-performance

Build timeout.

cheftako commented:
/retest

liggitt commented on pkg/controller/volume/persistentvolume/config/v1alpha1/defaults.go:
```diff
@@ -39,6 +39,11 @@ func RecommendedDefaultPersistentVolumeBinderControllerConfiguration(obj *kubect
 
+	if obj.VolumeHostAllowLocalLoopback == nil {
+		trueValue := true
+		obj.VolumeHostAllowLocalLoopback = &trueValue
```
is the intent to default this to true to match current behavior, pick that change back to 1.17/1.18, then flip the default to false for 1.20+?

mattcary replied:
Yes, I think so.

jsafrane above mentioned default-false would break some of his test setups, no-one was sure about real users. Out of an abundance of caution then we decided to keep it opt-in until we get something more definite.

I'm happy to hear another opinion, especially if it's more decisive!

liggitt commented on pkg/controller/volume/persistentvolume/config/types.go:
```diff
@@ -28,6 +28,13 @@ type PersistentVolumeBinderControllerConfiguration struct {
+	// VolumeHostCIDRDenylist is a list of CIDRs that should not be reachable by the
+	// controller from plugins.
+	VolumeHostCIDRDenylist []string
+	// VolumeHostAllowLocalLoopback indicates if local loopback hosts (127.0.0.1, etc)
+	// should be allowed when from plugins. This is a reference so that an unset value
+	// can be set for defaulting (see v1alpha1/defaults.go).
```
> VolumeHostAllowLocalLoopback indicates if local loopback hosts (127.0.0.1, etc) should be allowed when from plugins.

Sentence fragment?

> This is a reference so that an unset value can be set for defaulting (see v1alpha1/defaults.go).

I think this bit should be moved to v1alpha1/types.go? this var **isn't** a reference.

mattcary replied:
Arg, thank you.

liggitt commented on staging/src/k8s.io/kube-controller-manager/config/v1alpha1/types.go:
```diff
@@ -427,6 +427,12 @@ type PersistentVolumeBinderControllerConfiguration struct {
+	// VolumeHostCIDRDenylist is a list of CIDRs that should not be reachable by the
+	// controller from plugins.
+	VolumeHostCIDRDenylist []string
+	// VolumeHostAllowLocalLoopback indicates if local loopback hosts (127.0.0.1, etc)
+	// should be allowed when from plugins.
```
fragment?

mattcary replied:
Also fixed

mattcary commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

mattcary commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-integration

liggitt commented:
/approve
for flag/config changes

leaving lgtm to storage reviewers (@jsafrane @msau42)

msau42 commented:
lgtm, I just have a nit on the commits. the revert of quobyte vendoring is in the 2nd commit. Prefer that be merged into the 1st commit instead and 1st commit message updated to say that only storageos vendor got updated.

mattcary commented:
> lgtm, I just have a nit on the commits. the revert of quobyte vendoring is in the 2nd commit. Prefer that be merged into the 1st commit instead and 1st commit message updated to say that only storageos vendor got updated.

I've updated the commit comment.

I'm not sure what you mean by the quobyte vendoring being in the 2nd commit, AFAICT there's nothing in vendor in the second commit, that's all in the first commit. LMK what I've misunderstood.

msau42 commented:
> > lgtm, I just have a nit on the commits. the revert of quobyte vendoring is in the 2nd commit. Prefer that be merged into the 1st commit instead and 1st commit message updated to say that only storageos vendor got updated.
> 
> I've updated the commit comment.
> 
> I'm not sure what you mean by the quobyte vendoring being in the 2nd commit, AFAICT there's nothing in vendor in the second commit, that's all in the first commit. LMK what I've misunderstood.

Sorry I wasn't clear. The vendoring commit [updates quotbyte](https://github.com/kubernetes/kubernetes/pull/91785/commits/0fdfd7cf2899160fad0cd349a9d7f29b48622d11#diff-37aff102a57d3d7b797f152915a6dc16R87), and the other commit [reverts that vendor update](https://github.com/kubernetes/kubernetes/pull/91785/commits/82bff0b521f36d66d86918f6324fb5e912773d8e#diff-37aff102a57d3d7b797f152915a6dc16R87). It would be good to merge these two changes so that someone looking at an individual commit won't get confused.

mattcary commented:
> > > lgtm, I just have a nit on the commits. the revert of quobyte vendoring is in the 2nd commit. Prefer that be merged into the 1st commit instead and 1st commit message updated to say that only storageos vendor got updated.
> > 
> > 
> > I've updated the commit comment.
> > I'm not sure what you mean by the quobyte vendoring being in the 2nd commit, AFAICT there's nothing in vendor in the second commit, that's all in the first commit. LMK what I've misunderstood.
> 
> Sorry I wasn't clear. The vendoring commit [updates quotbyte](https://github.com/kubernetes/kubernetes/pull/91785/commits/0fdfd7cf2899160fad0cd349a9d7f29b48622d11#diff-37aff102a57d3d7b797f152915a6dc16R87), and the other commit [reverts that vendor update](https://github.com/kubernetes/kubernetes/pull/91785/commits/82bff0b521f36d66d86918f6324fb5e912773d8e#diff-37aff102a57d3d7b797f152915a6dc16R87). It would be good to merge these two changes so that someone looking at an individual commit won't get confused.

Ah, just in the go.mod. Gotcha.

mattcary commented:
/test pull-kubernetes-verify
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-integration
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-node-e2e

mattcary commented:
test pull-kubernetes-integration

msau42 commented:
/retest

msau42 commented:
/lgtm
/priority important-soon

msau42 commented:
Can you also update the PR description to remove gluster and quobyte, and add a note we'll have to update them separately

mattcary commented:
> Can you also update the PR description to remove gluster and quobyte, and add a note we'll have to update them separately

done

msau42 commented:
/lgtm

msau42 commented:
/milestone v1.20

The default behavior is unchanged, so this can be considered low risk.

justaugustus commented:
Looks like `/hack/update-vendor.sh` needs to be run on this one.

mattcary commented:
/test pull-kubernetes-verify

> Looks like `/hack/update-vendor.sh` needs to be run on this one.

neither udpate-bazal nor update-vendor doesn't change anything on my local client. verify-bazel passes locally for me. Maybe a flake?

mattcary commented:
Trying a rebase.

cheftako commented:
Looks like it needs ./hack/update-bazel.sh to be run again.

mattcary commented:
I've run update-bazel and verify-bazel many times after many different rebasing tries on two different machines from two different checkouts, and neither reports any updates or changes my commits. verify-bazel runs successfully. I've tried manually patching in the udpate-bazel output from the presubmit queue, but that just fails the bazel build.

I've tried various other hack/update-* or hack/verify-* and they report that everything is working.

I think I'll need to make a totally new kubernetes checkout with a new go install, but it will take me a little while to get the time to do that.

mattcary committed:
update-bazel

Change-Id: Iebc99ee13587f0cd4c43ab85c7295d458d679d1e

mattcary commented:
Hooray! update-bazel finally did something. Is it possible that one has to have a bazel-build succeed or somehow get sufficiently far before verify-bazel/update-bazel are effective? At any rate it seems happy now.

It made lots of updates in vendor/, I'm not sure if that's expected/desired?

msau42 commented:
/lgtm
Vendor changes seem reasonable, just storageos libraries

/retest
